### PR TITLE
Adding zeroing tensor for QuantizedCPU type in yaml

### PR DESCRIPTION
**Pull request by kwanghoon-meta**
Summary:
Gradient Checkpointing uses QUint8 type tensor quantizer which currently not supported in dispatch key.

By adding QuantizedCPU for zero_ operation, we enable gradient checkpointing compatible to custom backward function.

Since it is just filling zero element in a tensor, we can simply use same zero which uses CPU backend.

Test Plan: Signals

Differential Revision: D43144214

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#include <c10/core/ScalarType.h>
```
Comment (albanD) [inline aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:1]: Why is this include here and not with all the other ones?

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp [hunk 2]
```diff
@@ -65,13 +67,19 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
...
-        free(kernel);
-        return std::tuple<at::Tensor, c10::optional<at::Tensor>>(original_tensor, bias_);
+        float* weight_scales_data = w_scales.data_ptr<float>();
+        at::Tensor weight_origin;
+        weight_origin = at::empty(weight_sizes, at::device(c10::kCPU).dtype(at::kChar));
+        int8_t* weight_ptr_int8 =
+            reinterpret_cast<int8_t*>(weight_origin.data_ptr<int8_t>());
+        w->unpackWeights(w_zero_points.data(), weight_ptr_int8);
+        // See for the subtraction 128
+        // https://www.internalfb.com/code/fbsource/[D40927291-V8]/fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp?lines=326
```
Comment (albanD) [inline aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:77]: This should link to github code

File: buckbuild.bzl [hunk 3]
```diff
@@ -1909,7 +1943,146 @@ def define_buck_targets(
             C10,
         ],
     )
+    # Note : following aten_quantized_linear library is manually designed to identify all requires symbols to create a shared library
+    # This library is strictly required to add a feature quantized gradient function with backend kernel
+    # Because custom gradient function calls backend kernel matrix multiplication for gradient computation for memory efficiency
+    # ,which is located in aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp
+    # Refer - D42593235
```
Comment (albanD) [inline buckbuild.bzl:1950]: link PR

File: buckbuild.bzl [hunk 4]
```diff
@@ -1909,7 +1943,146 @@ def define_buck_targets(
...
+    aten_quantized_linear = [
```
Comment (albanD) [inline buckbuild.bzl:1951]: Why does this list need to be explicit here? I would expect we already have such a list somewhere?

File: buckbuild.bzl [hunk 5]
```diff
@@ -1483,6 +1484,39 @@ def define_buck_targets(
         ],
     )
 
+    pt_xplat_cxx_library(
```
Comment (albanD) [inline buckbuild.bzl:1487]: Why do you need these two new targets?

File: torch/csrc/quantized/quantized_backward.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+#endif
+
+namespace at {
+namespace native {
+namespace {
+#ifdef USE_PYTORCH_QNNPACK
+TORCH_LIBRARY_IMPL(quantized, Autograd, m) {
+  m.impl(
+      TORCH_SELECTIVE_NAME("quantized::linear_dynamic"),
```
Comment (albanD) [inline torch/csrc/quantized/quantized_backward.cpp:115]: Is this function in native_functions.yaml? If so all of this can just be codegened.

File: torch/csrc/quantized/quantized_backward.cpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    grad_fn = std::shared_ptr<PackedLinearWeightDynamicBackward>(
+        new PackedLinearWeightDynamicBackward(), deleteNode);
+    grad_fn->self_ = packed_weight;
+    grad_fn->input_sizes_ = input.sizes().vec();
+    grad_fn->set_next_edges(collect_next_edges(input));
+  }
+  static auto op = at::Dispatcher::singleton()
+                       .findSchemaOrThrow("quantized::linear_dynamic", "")
+                       .typed<decltype(packed_linear_weight_grad)>();
+  auto output = op.call(input, packed_weight, reduce_range);
```
Comment (albanD) [inline torch/csrc/quantized/quantized_backward.cpp:97]: This should be a redispatch.

File: torch/csrc/quantized/quantized_backward.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+#ifdef USE_PYTORCH_QNNPACK
+namespace caffe2 {
+CAFFE_KNOWN_TYPE(c10::intrusive_ptr<PackedLinearWeightsQnnp>);
+}
+namespace {
+using namespace torch::autograd;
+using namespace generated::details;
+using namespace at;
+// Gradient into argument 1 (`input`). TODO: Support bias as well.
+struct TORCH_API PackedLinearWeightDynamicBackward : public TraceableFunction {
```
Comment (albanD) [inline torch/csrc/quantized/quantized_backward.cpp:21]: You shouldn't write custom Node by hand. Use a custom Function to tie a fw and bw together: https://pytorch.org/tutorials/advanced/cpp_autograd.html#using-custom-autograd-function-in-c

In this case, the fw should just be the redispatch call like you do below.

Comment (kimishpatel) [reply torch/csrc/quantized/quantized_backward.cpp:21]: where is the fw call btw? I only see bw implementation here.

File: torch/csrc/quantized/quantized_backward.cpp (hunk 7 above)
Comment (kwanghoon-meta) [reply torch/csrc/quantized/quantized_backward.cpp:97]: I think this was last time I was suggested to use this way?, I'm unsure how to locate redispatch function for the scheme (can't find anywhere honestly)

File: torch/csrc/quantized/quantized_backward.cpp (hunk 6 above)
Comment (kwanghoon-meta) [reply torch/csrc/quantized/quantized_backward.cpp:115]: No, it is here https://www.internalfb.com/code/fbsource/[536eda60b18586b17be36889d4d09718ccf76d5b]/xplat/caffe2/aten/src/ATen/native/quantized/library.cpp?lines=153

File: torch/csrc/quantized/quantized_backward.cpp (hunk 8 above)
Comment (kwanghoon-meta) [reply torch/csrc/quantized/quantized_backward.cpp:21]: @albanD I'm honestly unsure if it is a feasible to implement this way, we lost an access to original weight with quantized tensor has a backend QNNPACK, it frees up weight tensor.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L335

File: torch/csrc/quantized/quantized_backward.cpp (hunk 6 above)
Comment (albanD) [reply torch/csrc/quantized/quantized_backward.cpp:115]: Ok ok sad....

File: torch/csrc/quantized/quantized_backward.cpp (hunk 7 above)
Comment (albanD) [reply torch/csrc/quantized/quantized_backward.cpp:97]: Yes in that link, it calls `op.redispatch` not `op.call`

File: torch/csrc/quantized/quantized_backward.cpp (hunk 8 above)
Comment (albanD) [reply torch/csrc/quantized/quantized_backward.cpp:21]: > where is the fw call btw?

The fw is in the backend key. The Autograd key is only responsible for setting up the graph.

Comment (albanD) [reply torch/csrc/quantized/quantized_backward.cpp:21]: > I'm honestly unsure if it is a feasible to implement this way, we lost an access to original weight tensor

I'm not sure what that changes? The signature of the custom Function is the same as packed_linear_weight_grad (you want to add the keyset to redispatch).
And from there you can save anything you need the same way you do with this one.
